### PR TITLE
Add Get-CVClientLicense/Revoke-CVClientLicense function

### PR DESCRIPTION
**Pull request by CraigTolley**
Hi,

Have added a new function to specifically get the client licenses. Am working on a counter function to remove client licenses as well.

Hope format is ok. Tried to keep with everything that you have already done. 

Craig

### CODE REVIEW

**Comment by CraigTolley** [comment]
Provides the license functionality in issue #4

File: Modules/Commvault.CommCell/Commvault.CommCell.psm1 [hunk 1]
```diff
@@ -1001,6 +1001,251 @@ function Set-CVClient {
...
+    Get-CVClientLicences -Name 'carbonwincs1'
+
+.OUTPUTS
+    [PsCustomObject] containing details of the license and the plaform type
+
+.NOTES
+    Author: Craig Tolley
+    Based on API Documentation at https://documentation.commvault.com/commvault/v11/article?p=47685.htm
+#>
+    [CmdletBinding(DefaultParameterSetName = 'Default')]
```
Comment (gstoops) [inline Modules/Commvault.CommCell/Commvault.CommCell.psm1]: default should be 'ByName'

File: Modules/Commvault.CommCell/Commvault.CommCell.psm1 [hunk 2]
```diff
@@ -1001,6 +1001,251 @@ function Set-CVClient {
...
+    [OutputType([PSCustomObject])]
+    param(
+        [Alias('Client')]
+        [Parameter(Mandatory = $False, ParameterSetName = 'ByName')]
```
Comment (gstoops) [inline Modules/Commvault.CommCell/Commvault.CommCell.psm1]: [Parameter(Mandatory = $True, ParameterSetName = 'ByName', ValueFromPipeline = $True, ValueFromPipelineByPropertyName = $True)]

File: Modules/Commvault.CommCell/Commvault.CommCell.psm1 [hunk 3]
```diff
@@ -1001,6 +1001,251 @@ function Set-CVClient {
...
+        [ValidateNotNullorEmpty()]
+        [String] $Name,
+
+        [Alias('ClientId')]
+        [Parameter(Mandatory = $False, ParameterSetName = 'ById')]
```
Comment (gstoops) [inline Modules/Commvault.CommCell/Commvault.CommCell.psm1]: [Parameter(Mandatory = $True, ParameterSetName = 'ByName', ValueFromPipeline = $True, ValueFromPipelineByPropertyName = $True)]

File: Modules/Commvault.CommCell/Commvault.CommCell.psm1 [hunk 4]
```diff
@@ -1001,6 +1001,251 @@ function Set-CVClient {
...
+            $sessionObj = Get-CVSessionDetail $MyInvocation.MyCommand.Name
+            $endpointSave = $sessionObj.requestProps.endpoint
+
+            if ($PSCmdlet.ParameterSetName -eq 'ByName' -or
+                $PSCmdlet.ParameterSetName -eq 'ById' ) {
+                $foundClient = $False
+            }
+            else {
+                $foundClient = $null
+            }
```
Comment (gstoops) [inline Modules/Commvault.CommCell/Commvault.CommCell.psm1]: Delete these lines.

File: Modules/Commvault.CommCell/Commvault.CommCell.psm1 [hunk 5]
```diff
@@ -1001,6 +1001,251 @@ function Set-CVClient {
...
+    }
+    
+    process { Write-Debug -Message "$($MyInvocation.MyCommand): process"
+
+        try {
+            $sessionObj.requestProps.endpoint = $endpointSave
+
+            if ($PSCmdlet.ParameterSetName -eq 'ByName') {
+                $Id = (Get-CVClient -Name $Name).clientId
+            }
```
Comment (gstoops) [inline Modules/Commvault.CommCell/Commvault.CommCell.psm1]: if ($PSCmdlet.ParameterSetName -eq 'ByName') {
                $clientObj = (Get-CVClient -Name $Name)
                if ($null -ne $clientObj) {
                    $Id = $clientObj.clientId
                }
                else {
                    Write-Information -InformationAction Continue -MessageData "INFO: $($MyInvocation.MyCommand): client not found having name [$Name]"
                    return
                }
            }

File: Modules/Commvault.CommCell/Commvault.CommCell.psm1 [hunk 6]
```diff
@@ -1001,6 +1001,251 @@ function Set-CVClient {
...
+
+            $sessionObj.requestProps.endpoint = $sessionObj.requestProps.endpoint -creplace ('{clientId}', $Id)
+            $headerObj = Get-CVRESTHeader $sessionObj
```
Comment (gstoops) [inline Modules/Commvault.CommCell/Commvault.CommCell.psm1:1454]: please add blank line here.

File: Modules/Commvault.CommCell/Commvault.CommCell.psm1 [hunk 7]
```diff
@@ -1001,6 +1001,251 @@ function Set-CVClient {
...
+            $payload = @{ }
+            $payload.Add('headerObject', $headerObj)
+            $payload.Add('body', $body)
+            $validate = 'licensesInfo'
+    
+            $response = Submit-CVRESTRequest $payload $validate
+    
+            if ($response.IsValid) {
+                Write-Output $response.Content.licensesInfo
+            }
```
Comment (gstoops) [inline Modules/Commvault.CommCell/Commvault.CommCell.psm1:1466]: if ($response.IsValid) {
                Write-Output $response.Content.licensesInfo
            }
            else {
                Write-Information -InformationAction Continue -MessageData "INFO: $($MyInvocation.MyCommand): license information not found for client Id [$Id]"
            }

File: Modules/Commvault.CommCell/Commvault.CommCell.psm1 [hunk 8]
```diff
@@ -1001,6 +1001,251 @@ function Set-CVClient {
...
+        [String] $Name,
+
+        [Alias('ClientId')]
+        [Parameter(Mandatory = $True, ParameterSetName = 'ById')]
+        [ValidateNotNullorEmpty()]
+        [Int32] $Id,
+
+        [Parameter(Mandatory = $True, ParameterSetName = 'ByName')]
+        [Parameter(Mandatory = $True, ParameterSetName = 'ById')]
+        $LicensesToRelease
```
Comment (gstoops) [inline Modules/Commvault.CommCell/Commvault.CommCell.psm1]: What type is this?

File: Modules/Commvault.CommCell/Commvault.CommCell.psm1 [hunk 9]
```diff
@@ -1001,6 +1001,251 @@ function Set-CVClient {
...
+
+.OUTPUTS
+    Outputs errorMessage and errorCode. Error code 0 is success.
+
+.NOTES
+    Author: Craig Tolley
+    Based on API Documentation at https://documentation.commvault.com/commvault/v11/article?p=92013.htm
+#>
+    Param (
+        [Alias('ClientName')]
```
Comment (gstoops) [inline Modules/Commvault.CommCell/Commvault.CommCell.psm1]: The $Name and $Id params here should be coded identical to the Get function

File: Modules/Commvault.CommCell/Commvault.CommCell.psm1 [hunk 10]
```diff
@@ -1001,6 +1001,251 @@ function Set-CVClient {
...
+    Revoke-CvClientLicenses -Name 'carbonwincs1' -LicensesToRelease $licenses
+
+.OUTPUTS
+    Outputs errorMessage and errorCode. Error code 0 is success.
+
+.NOTES
+    Author: Craig Tolley
+    Based on API Documentation at https://documentation.commvault.com/commvault/v11/article?p=92013.htm
+#>
+    Param (
```
Comment (gstoops) [inline Modules/Commvault.CommCell/Commvault.CommCell.psm1:1517]: [CmdletBinding(DefaultParameterSetName = 'ByName', SupportsShouldProcess = $True, ConfirmImpact = 'High')]

File: Modules/Commvault.RESTSession/Commvault.RESTSession.psm1 [hunk 11]
```diff
@@ -600,6 +599,22 @@ function GetAPIDetail ([String] $Request) {
...
+
+            'Revoke-CVClientLicense' = @{
+        
+                Description = 'Release a license from a client'
+                Endpoint    = 'Client/License/Release'
+                Method      = 'Post'
+                Body        = ''
+            
+            }
             'Add-CVClientGroup' = @{
```
Comment (gstoops) [inline Modules/Commvault.RESTSession/Commvault.RESTSession.psm1:638]: Please add blank line here.

**Review by gstoops — CHANGES_REQUESTED** [review]
Hi Craig, great work here. We just need to address a few things. :)
Please push your changes to Commvault.CommCell.psd1 for the new function exports:
FunctionsToExport = @('Get-CVAlert', 'Get-CVClientGroup', 'Get-CVClient', 'Get-CVClientLicense', 'Revoke-CVClientLicense',
                      'Get-CVSubclient', 'Get-CVVersionInfo', 'Set-CVClient', 'Set-CVClientGroup')

**Comment by CraigTolley** [comment]
All suggestions incorporated.

**Comment by CraigTolley** [comment]
Is there anything else that you need me to do before this can get merged in?

**Comment by CraigTolley** [comment]
Closed as updating my repos and creating a new PR

**Pull request closed without merging**